test-tools/addr2line/addr2line.py

```diff
@@ -96,29 +95,72 @@ def get_line_info(dwarf_dump: Path, wasm_file: Path, offset: int) -> str:
     )
     outputs = p.stdout.split(os.linesep)
 
-    capture_name = False
+    function_name, function_file = "<unknown>", "unknown"
+    function_line, function_column = "?", "?"
+
     for line in outputs:
         line = line.strip()
 
-        if "DW_TAG_subprogram" in line:
-            capture_name = True
-            continue
+        if "DW_AT_name" in line:
+            function_name = get_dwarf_tag_value("DW_AT_name", line)
+
+        if "DW_AT_decl_file" in line:
+            function_file = get_dwarf_tag_value("DW_AT_decl_file", line)
+
+        if "Line info" in line:
+            _, function_line, function_column = parse_line_info(line)
+
+    return (function_name, function_file, function_line, function_column)
+
+
+def get_dwarf_tag_value(tag: str, line: str) -> str:
+    # Try extracting value as string
+    STR_PATTERN = rf"{tag}\s+\(\"(.*)\"\)"
+    m = re.match(STR_PATTERN, line)
+    if m:
+        return m.groups()[0]
+
+    # Try extracting value as integer
+    INT_PATTERN = rf"{tag}\s+\((\d+)\)"
+    m = re.match(INT_PATTERN, line)
+    return m.groups()[0]
+
+
+def get_line_info_from_function_name(
+    dwarf_dump: Path, wasm_file: Path, function_name: str
+) -> tuple[str, str, str]:
+    """
+    Find the location info of a given function in a wasm file.
+    """
+    cmd = f"{dwarf_dump} --name={function_name} {wasm_file}"
+    p = subprocess.run(
+        shlex.split(cmd),
+        check=False,
+        capture_output=True,
+        text=True,
+        universal_newlines=True,
+    )
+    outputs = p.stdout.split(os.linesep)
+
+    function_name, function_file = "<unknown>", "unknown"
+    function_line = "?"
+
+    for line in outputs:
+        line = line.strip()
 
-        if "DW_AT_name" in line and capture_name:
-            PATTERN = r"DW_AT_name\s+\(\"(\S+)\"\)"
-            m = re.match(PATTERN, line)
-            assert m is not None
+        if "DW_AT_name" in line:
+            function_name = get_dwarf_tag_value("DW_AT_name", line)
 
-            function_name = m.groups()[0]
+        if "DW_AT_decl_file" in line:
+            function_file = get_dwarf_tag_value("DW_AT_decl_file", line)
 
-        if line.startswith("Line info"):
-            location = line
-            return (function_name, location)
+        if "DW_AT_decl_line" in line:
+            function_line = get_dwarf_tag_value("DW_AT_decl_line", line)
```

Comment:
duplicated with L154 and L155 ?

Reply:
What do you mean? L154-155 are for the file, L157-158 for the line